### ISSUE TITLE
concat: sharded concat above 16 inputs asserts instead of batching, and batching cannot work with a shared output_mem_config

### ISSUE THREAD

**Issue by VirdhatchaniKN** [tenstorrent/tt-metal#56824]
Found while reviewing #56603. Independent of that PR's height-concat layout bug — this is the input-count path and affects width and height sharded concat alike.

## Problem

`ConcatS2SMultiProgramFactory` aliases each input shard directly as a circular buffer, indices `0..N-1` with the output at 16, so it hard-asserts above 16 inputs:

```
concat_s2s_multi_program_factory.cpp:50
TT_FATAL(num_input_tensors <= cb_dst_id, "Not enough circular buffer for {} inputs.", ...)
```

But `calculate_max_tensors_per_concat` advertises a much higher ceiling for width/height sharded — derived from the kernel's runtime-arg budget, currently `(256-6)/5 = 50`, `x0.9 = 45`. Since 45 > 16, `concat_impl` does not batch anything in the range 17..45: those inputs go straight to the factory and hit the CB assert.

**Capping the advertised limit at 16 does not fix it.** The batching path hands every batch the *final* `output_mem_config`:

```
concat_device_operation.cpp:424
Tensor batch_result = concat_impl(batch, dim, groups, output_mem_config, sub_core_grids);
```

That config's shard height is sized for all N inputs, so a 16-input batch produces 128 rows while being asked for a 136-row shard, and fails in `TensorSpec`. Batching for sharded concat cannot work while every batch reuses the final output config — which is presumably why the block-sharded path rejects batching outright at `concat_device_operation.cpp:398` ("Batching with intermediate shard specs is not supported for block-sharded concat"). Width/height sharded has the same limitation with no equivalent guard.

## Hardware-confirmed (Wormhole)

Width-sharded height concat, `shape=(2, 4, 16)` per input, `dim=-2`, 1 core, uint32, ROW_MAJOR:

| inputs | result |
|--|--|
| 2 | correct |
| 16 | correct |
| 17, 20, 33, 40 | `TT_FATAL ... concat_s2s_multi_program_factory.cpp:50: Not enough circular buffer for N inputs` |

With the advertised limit locally capped to 16 so batching engages, 17 inputs instead gives:

```
TT_FATAL ... tensor_spec.cpp:162: Shard height 136 must match physical height 128 for width sharded
```

136 is the shard sized for all 17 inputs; 128 is what the first batch of 16 actually produces.

## Repro

```python
import torch, ttnn
dev = ttnn.open_device(device_id=0)
grid = ttnn.CoreRangeSet({ttnn.CoreRange(ttnn.CoreCoord(0, 0), ttnn.CoreCoord(0, 0))})
mk = lambda h: ttnn.create_sharded_memory_config(
    (h, 16), core_grid=grid, strategy=ttnn.ShardStrategy.WIDTH,
    orientation=ttnn.ShardOrientation.ROW_MAJOR, use_height_and_width_as_shard_shape=True)

for n in (16, 17):
    shape, flat = (2, 4, 16), 8
    ins = [(torch.arange(flat * 16, dtype=torch.int64) + i * 100000).to(torch.int32).reshape(shape)
           for i in range(n)]
    tt = [ttnn.to_memory_config(
              ttnn.from_torch(x, layout=ttnn.ROW_MAJOR_LAYOUT, device=dev, dtype=ttnn.uint32), mk(flat))
          for x in ins]
    out = ttnn.concat(tt, dim=-2, memory_config=mk(flat * n))   # n=17 asserts
    assert torch.equal(torch.concat(ins, dim=-2).to(torch.int64), ttnn.to_torch(out).to(torch.int64))
```

## Impact

Hard error, not wrong data — so this is a capability gap and a diagnosability gap, not a correctness bug. Any caller concatenating more than 16 sharded tensors fails, and the advertised limit of 45 is misleading to anyone reading it as the real ceiling.

## Suggested fix

Two parts, in order:

1. **Give each batch its own output config.** Batching has to derive a per-batch `MemoryConfig` whose shard height matches that batch's input count, instead of reusing the final one. The final concat over the intermediates then uses the caller's config as today.
2. **Then cap the sharded advertised limit at the CB ceiling (16)**, so `concat_impl` actually batches at the point the factory can no longer take more inputs.

If per-batch shard specs are not wanted, the smaller alternative is to mirror the block-sharded guard: reject more than 16 width/height sharded inputs up front with a message naming the CB ceiling, so the failure is at least self-explanatory. That is strictly a diagnosability fix and leaves the capability gap.

Note the ordering matters — doing (2) without (1) is a regression in diagnosability, turning a clear `Not enough circular buffer for 17 inputs` into an opaque shard-height mismatch three layers away. Measured, not hypothesised.

## Test plan

- Width- and height-sharded concat at 2, 16, 17, 33 and 64 inputs, comparing values and shape against `torch.concat` with `assert_equal`
- Both `dim == rank-1` and `dim == rank-2`, since the batching intermediate's shard spec differs per direction
- A leading dim > 1, so batching composes with the per-leading-index interleaving from #55342
- Block-sharded stays rejected with its existing message
